batchflow/models/torch/base.py

```diff
@@ -631,36 +632,46 @@ def _make_optimizer(self, config):
         else:
             raise ValueError("Optimizer is not defined", optimizer)
 
-        decay, decay_args = self._make_decay(config)
-        if decay is not None:
-            decay = decay(optimizer, **decay_args)
-        return optimizer, decay
+        decays, list_kwargs, list_steps = self._make_decay(config)
+
+        if decays:
+            decays = [decay(optimizer, **decay_args) for decay, decay_args in zip(decays, list_kwargs)]
+        return optimizer, decays, list_steps
 
     def _make_decay(self, config):
-        decay, decay_args = unpack_fn_from_config('decay', config)
-        n_iters = config.get('n_iters')
+        res = unpack_fn_from_config('decay', config)
+        res = res if isinstance(res, list) else [res]
+        decays, list_kwargs, list_steps = [], [], []
+        for decay, decay_args in res:
+            if decay is None:
+                return decays, list_kwargs, list_steps
+
+            step_meta_keys = ['frequency', 'first_iter', 'last_iter']
+            step_meta_defaults = [None, 0, np.finfo(np.float).max]
+            step_meta = {key: decay_args.pop(key, default) for key, default in zip(step_meta_keys, step_meta_defaults)}
+
+            if step_meta['frequency'] is None:
+                raise ValueError("Missing required key 'decay/frequency' in the configuration dict.")
+            if callable(decay) or isinstance(decay, type):
+                pass
+            elif isinstance(decay, str) and hasattr(torch.optim.lr_scheduler, decay):
+                decay = getattr(torch.optim.lr_scheduler, decay)
+            elif decay in DECAYS:
+                decay = DECAYS.get(decay)
+            else:
+                raise ValueError("Unknown learning rate decay method", decay)
 
-        if decay is None:
-            return decay, decay_args
-        if 'n_iters' not in config:
-            raise ValueError("Missing required key ``'n_iters'`` in the cofiguration dict.")
+            if decay in DECAYS_DEFAULTS:
+                decay_dict = DECAYS_DEFAULTS.get(decay).copy()
+                if decay == DECAYS['cos']:
+                    decay_dict.update(T_max=step_meta['frequency'])
+                decay_dict.update(decay_args)
+                decay_args = decay_dict.copy()
```

Comment:
Why do you need this line at all, when you already have `decay_dict`? And, btw, since it's a dict, calling variable `'smth_args'` is incorrect and misleading.

Reply:
All of it needs for replacing default decay values. And in the end, only `decay_args` is saved.